File: src/main/c/jni-lib/src/events/SummaryMapping.cpp
// SPDX-License-Identifier: MPL-2.0

#include "dxfeed/events/SummaryMapping.hpp"
#include "dxfeed/utils/ByteReader.hpp"
#include "dxfeed/utils/ByteWriter.hpp"

namespace dxfeed::jni {
  dxfg_summary_t* SummaryMapping::toSummary(ByteReader& reader) {
    auto summary = new dxfg_summary_t();
    summary->market_event.event_type.clazz = DXFG_EVENT_SUMMARY;
    summary->market_event.event_symbol = reader.readString();
    summary->market_event.event_time = reader.readLong();
    summary->day_id = reader.readInt();
    summary->prev_day_id = reader.readInt();
    summary->open_interest = reader.readLong();
    summary->flags = reader.readInt();

    summary->day_open_price = reader.readDouble();
    summary->day_high_price = reader.readDouble();
    summary->day_low_price = reader.readDouble();
    summary->day_close_price = reader.readDouble();
    summary->prev_day_close_price = reader.readDouble();
    summary->prev_day_volume = reader.readDouble();
    return summary;
  }

  void SummaryMapping::fromSummary(dxfg_summary_t* summary, ByteWriter& writer) {
    writer.writeString(summary->market_event.event_symbol);
    writer.writeInt64_t(summary->market_event.event_time);
    writer.writeInt32_t(summary->day_id);
    writer.writeInt32_t(summary->prev_day_id);
    writer.writeInt64_t(summary->open_interest);
    writer.writeInt32_t(summary->flags);

    writer.writeDouble(summary->day_open_price);
    writer.writeDouble(summary->day_high_price);
    writer.writeDouble(summary->day_low_price);
    writer.writeDouble(summary->day_close_price);
    writer.writeDouble(summary->prev_day_close_price);
    writer.writeDouble(summary->prev_day_volume);
  }
}

File: src/main/java/com/dxfeed/api/serializers/SummaryMapping.java

package com.dxfeed.api.serializers;

import com.dxfeed.api.buffers.ByteBuffer;
import com.dxfeed.api.buffers.DoubleBuffer;
import com.dxfeed.api.buffers.ByteReader;
import com.dxfeed.event.market.DxFeedEventMarketPackagePrivate;
import com.dxfeed.event.market.Summary;

public class SummaryMapping {
  /**
   * typedef struct dxfg_market_event_t {
   *    dxfg_event_type_t event_type;
   *    const char *event_symbol;
   *    int64_t event_time;
   * } dxfg_market_event_t;
   *
   * https://github.com/dxFeed/dxfeed-graal-native-sdk/blob/main/src/main/c/api/dxfg_events.h#L182
   *
   * typedef struct dxfg_summary_t {
   *    dxfg_market_event_t market_event;
   *    //    dxfg_lasting_event_t lasting_event;
   *    int32_t day_id;
   *    double day_open_price;
   *    double day_high_price;
   *    double day_low_price;
   *    double day_close_price;
   *    int32_t prev_day_id;
   *    double prev_day_close_price;
   *    double prev_day_volume;
   *    int64_t open_interest;
   *    int32_t flags;
   * } dxfg_summary_t;
   */

  public static void toNative(Summary event, ByteBuffer pBytes, DoubleBuffer pDoubles) {
    // BYTE DATA
    pBytes.writeString(event.getEventSymbol());                       // 2 + eventSymbolLength
    pBytes.writeLong(event.getEventTime());                           // 8
    pBytes.writeInt(event.getDayId());                                // 4
    pBytes.writeInt(event.getPrevDayId());                            // 4
    pBytes.writeLong(event.getOpenInterest());                        // 8
    pBytes.writeInt(DxFeedEventMarketPackagePrivate.getFlags(event)); // 4

    // DOUBLE DATA
    pDoubles.write(event.getDayOpenPrice());
    pDoubles.write(event.getDayHighPrice());
    pDoubles.write(event.getDayLowPrice());
    pDoubles.write(event.getDayClosePrice());
    pDoubles.write(event.getPrevDayClosePrice());
    pDoubles.write(event.getPrevDayVolume());
  }

  public static Summary fromNative(ByteReader reader) {
    Summary summary = new Summary();
    summary.setEventSymbol(reader.readString());
    summary.setEventTime(reader.readLong());
    summary.setDayId(reader.readInt());
    summary.setPrevDayId(reader.readInt());
    summary.setOpenInterest(reader.readLong());
    DxFeedEventMarketPackagePrivate.setFlags(summary, reader.readInt());

    summary.setDayOpenPrice(reader.readDouble());
    summary.setDayHighPrice(reader.readDouble());
    summary.setDayLowPrice(reader.readDouble());
    summary.setDayClosePrice(reader.readDouble());
    summary.setPrevDayClosePrice(reader.readDouble());
    summary.setPrevDayVolume(reader.readDouble());
    return summary;
  }
}
